config/ompi_setup_fc.m4

```diff
@@ -50,9 +50,168 @@ AC_DEFUN_ONCE([_OMPI_SETUP_FC_COMPILER],[
+dnl   certain versions of Xcode.  In these failure scenarios, we add
+dnl   -Wl,-ld_classic to make it work.
+dnl
+dnl Note: the -ld_classic flag works with both static and shared
+dnl libraries, so it is safe to always do this test -- even if
+dnl we are only building static libraries.
+dnl
+dnl The history is that for a long time (decades),
+dnl -Wl,-commons,use_dylibs worked by itself.  But then Apple decided
+dnl to remove support for it, and cause link-time errors with MPI
+dnl Fortran sentinel values (e.g., MPI_BOTTOM) because they are
+dnl implemented with Fortran common blocks.  Later, Xcode v16
+dnl (re)added support for -Wl,commons,use_dylibs again.  Hence, there
+dnl is a window of a few versions of Xcode that need the additional
+dnl -Wl,ld_classic flag to link/work properly.
+dnl
+dnl We have to use a slightly complex test code that will actually
+dnl fail if the version of Xcode being used requires -Wl,-ld_classic
+dnl with -Wl,-commons,use_dylibs.
+dnl
+dnl 1. Build a shared library (with C source code) with a public
+dnl    symbol that can be used as a Fortran common block symbol.
+dnl 2. Compile a Fortran program that calls a function in the shared
+dnl    library, and link it against the shared library.
+dnl
+dnl Note: This is a linker test; we are checking to see if this all
+dnl compiles and links properly.  The logic in the C / Fortran code
+dnl below specifically does not test for correctness because we do not
+dnl actually run the code.
+AC_DEFUN([_OMPI_SETUP_FC_XCODE_COMMONS_LDFLAGS],[
+    OPAL_VAR_SCOPE_PUSH([xcode_flags])
+
+    # This variable is used by the invoking macro to display the
+    # results via AC RESULT (just to keep the symmetry of
+    # MSG_CHECKING / RESULT in the same upper-level macro).
+    OMPI_FORTRAN_WRAPPER_FLAGS=
+
+    xcode_flags="-Wl,-commons,use_dylibs"
+    _OMPI_SETUP_FC_XCODE_COMMONS_LDFLAGS_BACKEND(
+        [$xcode_flags],
+        [OMPI_FORTRAN_WRAPPER_FLAGS=$xcode_flags], [])
+    AS_IF([test -z "$OMPI_FORTRAN_WRAPPER_FLAGS"],
+          [xcode_flags="-Wl,-commons,use_dylibs -Wl,-ld_classic"
+           _OMPI_SETUP_FC_XCODE_COMMONS_LDFLAGS_BACKEND(
+               [$xcode_flags],
+               [OMPI_FORTRAN_WRAPPER_FLAGS=$xcode_flags], [])])
+    AS_IF([test -z "$OMPI_FORTRAN_WRAPPER_FLAGS"],
+          [OMPI_FORTRAN_WRAPPER_FLAGS="none"])
+
+    OPAL_VAR_SCOPE_POP
+])
+
+dnl Companion to _OMPI SETUP_FC_XCODE_COMMONS_LDFLAGS;
+dnl see that macro for an explanation of this macro.
+dnl
+dnl $1: LDFLAGS to test
+dnl $2: action to perform upon success
+dnl $3: action to perform upon failure
+AC_DEFUN([_OMPI_SETUP_FC_XCODE_COMMONS_LDFLAGS_BACKEND],[
+    OPAL_VAR_SCOPE_PUSH([xcode_happy xcode_dir LDFLAGS_save_xcode LIBS_save_xcode])
+
+    xcode_dir=conftest.$$
+    rm -rf $xcode_dir
+    mkdir -p $xcode_dir
+    cd $xcode_dir
+
+    LIBS_save_xcode=$LIBS
+    LDFLAGS_save_xcode=$LDFLAGS
+    LDFLAGS="$LDFLAGS -L. $1"
```

Comment:
Why are you dropping all of LDFLAGS?  We wouldn't catch conflicting flags on picking the linker version (`-ld_new`) or incompatible options (`-merge*`).

Reply:
@bwbarrett and I talked this through IRL: all is ok.